### PR TITLE
Containerized code: double quotes setting of computer and code setup

### PR DESCRIPTION
**Pull request by unkcpz**
The PR is a part required for containerized code implementation (also for hyperqueue without hacky workaround to evaluate command-line variable by `$` properly) and is separated from https://github.com/aiidateam/aiida-core/pull/5250. It allows users to set up computer and code with an extra option `--use-double-quotes` to using double quotes for the command run line of job script.

### CODE REVIEW

**Comment by ltalirz** [comment]
Thanks @unkcpz . At a high level, I'm just wondering: do we really need to push this choice to the user?

I.e. can't we be clever about using double quotes where necessary (if we detect some environment variable / ...)? This will be yet another option in the computer an code setup and I wonder whether people really need to care about this.

**Comment by unkcpz** [comment]
Hi @ltalirz, yes ideally it should be intelligent enough to know when to use double quotes and not. But I think it is hard to decide by what criteria we escape it with double or single quotes. Setting it manually might be a safe way to keep the backward compatibility.

File: aiida/schedulers/datastructures.py [hunk 1]
```diff
@@ -374,7 +374,9 @@ class JobTemplateCodeInfo:
     :param stderr_name: filename of the the `stderr` file descriptor.
     :param join_files: boolean, if true, `stderr` should be redirected to `stdout`.
     """
+    prepend_cmdline_params: list[str] = field(default_factory=list)
     cmdline_params: list[str] = field(default_factory=list)
+    use_double_quotes: list[bool] = field(default_factory=list)
```
Comment (sphuber) [inline aiida/schedulers/datastructures.py:379]: Please update the docstring:
```python
:param preprend_cmdline_params: list of command line arguments that are to be prepended to the executable.
:param use_double_quotes: list of two booleans. If true, use double quotes to escape command line arguments. The first
    value applies to `prepend_cmdline_params` and the second to `cmdline_params`.
```
What is `prepend_cmdine_params

File: aiida/engine/processes/calcjobs/calcjob.py [hunk 2]
```diff
@@ -708,16 +708,17 @@ def presubmit(self, folder: Folder) -> CalcInfo:
...
-                    (code_info.cmdline_params if code_info.cmdline_params is not None else [])
-                )
+                prepend_cmdline_params = mpi_args + extra_mpirun_params
             else:
-                this_argv = [this_code.get_execname()
-                             ] + (code_info.cmdline_params if code_info.cmdline_params is not None else [])
+                prepend_cmdline_params = []
+
+            cmdline_params = [this_code.get_execname()
+                              ] + (code_info.cmdline_params if code_info.cmdline_params is not None else [])
```
Comment (sphuber) [inline aiida/engine/processes/calcjobs/calcjob.py:716]:
```suggestion
            cmdline_params = [this_code.get_execname()] + (code_info.cmdline_params or [])
```

File: aiida/cmdline/params/options/commands/computer.py [hunk 3]
```diff
@@ -117,6 +117,16 @@ def should_call_default_mpiprocs_per_machine(ctx):  # pylint: disable=invalid-na
     help='The default amount of RAM (kB) that should be allocated per machine (node), if not otherwise specified.'
 )
 
+USE_DOUBLE_QUOTES = OverridableOption(
+    '--use-double-quotes/--not-use-double-quotes',
+    is_eager=False,
```
Comment (sphuber) [inline aiida/cmdline/params/options/commands/computer.py:122]: This is the default and is not necessary
```suggestion
    is_eager=False,
```

File: aiida/cmdline/params/options/commands/code.py [hunk 4]
```diff
@@ -109,6 +109,15 @@ def validate_label_uniqueness(ctx, _, value):
     help='[if --store-in-db]: Relative path of the executable inside the code-folder.'
 )
 
+USE_DOUBLE_QUOTES = OverridableOption(
+    '--use-double-quotes/--not-use-double-quotes',
+    is_eager=False,
```
Comment (sphuber) [inline aiida/cmdline/params/options/commands/code.py:114]:
```suggestion
```

File: aiida/cmdline/params/options/commands/code.py [hunk 5]
```diff
@@ -109,6 +109,15 @@ def validate_label_uniqueness(ctx, _, value):
...
+    default=False,
+    cls=InteractiveOption,
+    prompt='Whether use double quotes for code cmdline params?',
```
Comment (sphuber) [inline aiida/cmdline/params/options/commands/code.py:117]:
```suggestion
    prompt='Escape CLI arguments in double quotes',
```

File: aiida/cmdline/params/options/commands/computer.py [hunk 6]
```diff
@@ -117,6 +117,16 @@ def should_call_default_mpiprocs_per_machine(ctx):  # pylint: disable=invalid-na
...
+    default=False,
+    cls=InteractiveOption,
+    prompt='Whether use double quotes for prepend cmdline params?',
```
Comment (sphuber) [inline aiida/cmdline/params/options/commands/computer.py:125]:
```suggestion
    prompt='Escape CLI arguments in double quotes',
```

File: aiida/cmdline/params/options/commands/code.py [hunk 7]
```diff
@@ -109,6 +109,15 @@ def validate_label_uniqueness(ctx, _, value):
...
+    help='Whether the code executable parameters in script using the double quotes to escape.'
```
Comment (sphuber) [inline aiida/cmdline/params/options/commands/code.py:118]:
```suggestion
    help='Whether the executable and arguments of the code in the submission script should be escaped with single or double quotes.'
```

File: aiida/cmdline/params/options/commands/computer.py [hunk 8]
```diff
@@ -117,6 +117,16 @@ def should_call_default_mpiprocs_per_machine(ctx):  # pylint: disable=invalid-na
...
+    help='Whether the cmdline parameters before the code executable parameters in '
+    'script using the double quotes to escape.',
```
Comment (sphuber) [inline aiida/cmdline/params/options/commands/computer.py:127]:
```suggestion
    help='Whether the command line arguments before the executable in the submission script should be escaped with single or double quotes.'
```

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @unkcpz . Mostly minor comments, but one fundamental question: is the separate control of quoting in the `Code` and `Computer` necessary? And if so, why should the quoting for the `Computer` only apply to the argument before the executable and not any that follows, such as the input and output redirection?

**Review by unkcpz — COMMENTED** [review]
@sphuber sorry for the delay of changes and reply, suffer from COVID last week. 

I implement most of your suggestions and also docstring of `use_double_quotes` of `JobTemplateCodeInfo`. 

> but one fundamental question: is the separate control of quoting in the Code and Computer necessary?

I think there are two reason to apply it independently: 1. it is more flexible and more possible for backward compatible if it can be controlled separately. 2. with initial containerized code implementation (the PR after this PR) the specific suffix for using containerized code is independent of `cmdline_params` parts of code and double quotes escape needed since I want to use `$PWD` to map working dir to directory where the inputs locate. While for the code part it can keep the same.

File: aiida/orm/nodes/data/code.py [hunk 9]
```diff
@@ -351,6 +351,20 @@ def get_input_plugin_name(self):
...
+        Set whether a code cmdline is escape with double quotes.
+        """
+        from aiida.common.lang import type_check
+        type_check(use_double_quotes, bool)
+        self.base.attributes.set('use_double_quotes', use_double_quotes)
+
+    def get_use_double_quotes(self):
+        """
+        Return whether a code is escape with double quotes (False, default) or not (True).
+        """
```
[the remark this answers is not in the document]
Comment (sphuber) [reply aiida/orm/nodes/data/code.py:365]: Same here. Don't see the changes.

**Comment by sphuber** [comment]
Thanks @unkcpz . Hope you are feeling better. I noticed that you resolved a few comments that don't seem to actually have been addressed. At least I cannot see the changes of the suggestions I made. Also the tests and pre-commit are failing. Please fix those and update the branch.

**Comment by unkcpz** [comment]
@sphuber sorry for the carelessness, I wrongly mix up commits and push to the actual container code implementation branch locally previously. Now it is rebased a bit and passes all the tests.

**Review by sphuber — APPROVED** [review]

**Comment by sphuber** [comment]
Thanks @unkcpz !

**Pull request merged**